### PR TITLE
Add ScopedSetupWithInputs (stacked 2/3)

### PR DESCRIPTION
**Pull request by AntonLydike**
This PR adds ScopedSetupWithInputs and moves a bunch of helpers out of the rewrites and into the `inference` module.

ScopedSetupWithInputs represents a setup op together with operations that calculate it's inputs. These operations must be pure, and be contained within a block (usually the block at which the operation is located).

This helper class is needed for stacked PR 3/3, and will also be helpful to simplify some other rewrites I'm sure,

### CODE REVIEW

File: compiler/inference/helpers.py
Comment (AntonLydike) [inline compiler/inference/helpers.py:1]: Nothin in this file is new, it's just stuff moved out of the specific rewrites in an attempt to clean up.

File: compiler/inference/helpers.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    if end_op is not None:
+        assert (
+            start_op.parent_block() is end_op.parent_block()
+        ), "Cannot iterate between operations not within the same block!"
+
+    yield start_op
+    while start_op.next_op is not end_op:
+        assert start_op.next_op is not None  # just for pyright
+        yield start_op.next_op
+        start_op = start_op.next_op
```
Comment (AntonLydike) [inline compiler/inference/helpers.py:141]: This is a neat helper that I might upstream to xDSL at some point. It's new in this PR

File: compiler/transforms/accfg_config_overlap.py [hunk 2]
```diff
@@ -53,10 +57,30 @@ def match_and_rewrite(self, op: accfg.SetupOp, rewriter: PatternRewriter, /):
...
+        if all(
+            between_op in inputs.inputs for between_op in iter_ops_range(launch, op)
+        ):
+            return
+
+        inputs.lazy_move_up(  # and move the setup and inputs to be right behind the launch
+            parent_block,
+            InsertPoint.after(launch),
+            rewriter,
+        )
```
Comment (JosseVanDelm) [inline compiler/transforms/accfg_config_overlap.py:83]: Maybe we should make clear that this is a heuristic? We can move these ops anywhere in the block iiuc, and not all possible move strategies might make as much sense depending on the scenario?

**Review by JosseVanDelm — APPROVED** [review]
I've added a few comments:
It seems like this is implemented for scf.for ops (in the comments), but the code seems to cover more stuff? Can you clarify/address those comments?
Apart from that this looks very neat!

Can you have a look at my heuristic comment and tell me what you think about it?

**Comment by AntonLydike** [comment]
Thanks for the feedback!

File: compiler/transforms/accfg_config_overlap.py (hunk 2 above)
Comment (AntonLydike) [reply compiler/transforms/accfg_config_overlap.py:83]: Not sure what you mean exactly by this...

**Pull request merged**